doc/TockBinaryFormat.md

```diff
@@ -47,54 +61,66 @@ understand important aspects of the app. In particular, the kernel must know
 
-Apps in Tock create an effective linked list structure in flash. That is, the
-start of the next app is immediately at the end of the previous app. Therefore,
-the TBF header must specify the length of the app so that the kernel can find
-the start of the next app.
+## App Storage
 
-If there is a gap between apps an "empty app" can be inserted to keep the linked
-list structure intact.
+TBF Objects in Tock are stored sequentially.  The start of TBF Object
+N+1 is immediately at the end of TBF Object N. Therefore, the TBF
+header specifies the length of the TBF Object so that the kernel can
+find the start of the next one.
 
-Also, functionally Tock apps are sorted by size from longest to shortest. This
-is to match MPU rules about alignment.
+If there is a gap between TBF Objects an "empty object" can be
+inserted to keep the structure intact.
+
+Tock apps are typically stored in sorted order, from longest to
+shortest. This is to help match MPU rules about alignment.
 
 ## Empty Tock Apps
 
-An "app" need not contain any code. An app can be marked as disabled and
-effectively act as padding between apps.
+A TBF Object can contain no code. A TBF Object can be marked as
+disabled to be padding between other objects.
+
+## TBF Header Section
 
-## TBF Header
+The TBF Header section contains all of a TBF Object's headers. All TBF
+Objects have a Base Header and the Base Header is always first.  All
+headers are a multiple of 4 bytes long; the TBF Header section is
+multiple of 4 bytes long.
 
-The fields of the TBF header are as shown below. All fields in the header are
-little-endian.
+These are the Rust structures the kernel uses, defined in the
+`tock-tbf` crate, to represent headers.  Their in-flash
+representations are described below.
 
 ```rust
-struct TbfHeader {
+struct TbfHeaderV2Base {
     version: u16,            // Version of the Tock Binary Format (currently 2)
-    header_size: u16,        // Number of bytes in the complete TBF header
+    header_size: u16,        // Number of bytes in the TBF header section
     total_size: u32,         // Total padded size of the program image in bytes, including header
     flags: u32,              // Various flags associated with the application
     checksum: u32,           // XOR of all 4 byte words in the header, including existing optional structs
-
-    // Optional structs. All optional structs start on a 4-byte boundary.
-    main: Option<TbfHeaderMain>,
-    pic_options: Option<TbfHeaderPicOption1Fields>,
-    name: Option<TbfHeaderPackageName>,
-    flash_regions: Option<TbfHeaderWriteableFlashRegions>,
-    fixed_address: Option<TbfHeaderV2FixedAddresses>,
-    permissions: Option<TbfHeaderV2Permissions>,
-    persistent_acl: Option<TbfHeaderV2PersistentAcl>,
 }
+```
+
+After the Base Header come optional headers. Optional headers are
+structured as TLVs (type-length-values). Footers are encoded in the
+same way. Footers are also called headers for historical reasons.
```

Comment:
> Footers are also called headers for historical reasons.

What does this mean? Is this saying there are things in the `enum` below called `TbfHeaderXX` that really belong in the footer? (The type _numbers_ are public API, but the names could be switched no?)

Or is this saying there are types that can go _either_ in the header or footer (i.e. depending on whether integrity is important)?

---

Update: I was curious if there was a unifying word for the area that headers and footers go into: https://english.stackexchange.com/questions/24060/what-word-defines-a-category-suited-for-both-header-and-footer -- seems like no 😆 

---

Still, should we rename these to something like `TbfObjectType` to be more consistent with the language earlier, and reserve `TbfHeaderType` and `TbfFooterType` for things that can only go in the former or latter?

Reply:
What this means is there were originally only headers, and using the same infrastructure and naming to also cover types after the binary keeps things backwards compatible without introducing additional complexity.